packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,115 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL } from './constants';
+import * as React from 'react';
+import { SlotPropsRecord } from './types';
+import { ExtractSlotProps } from './types';
+import { PropsWithoutRef } from './types';
+import { getNativeElementProps } from '../utils/getNativeElementProps';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  elementType?:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+};
+
+/**
+ * Creates a slot for a v9 component
+ * @param params - an object containing:
+ * 1. `value` provided through `props.SLOT`,
+ * 2. Base `elementType` that will be rendered
+ * 3. A boolean (`required`) indicating if the slot will be rendered even if no `value` is provided
+ * 4. `defaultProps`, similar to a component a slot might have default properties to be merged together with the value provided by the user
+ *
+ * @example
+ * ```ts
+ * // useDialogSurface
+ * const backdropSlot = slot({
+    value: props.backdrop,
+    elementType: 'div',
+    required: open && modalType !== 'non-modal',
+    defaultProps: { 'aria-hidden': 'true' },
+  });
+ * ```
+ */
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | SlotComponent<Props> | undefined,
+  options: { required: true } & SlotOptions<Props>,
+): SlotComponent<Props>;
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | SlotComponent<Props> | undefined | null,
+  options?: { required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | SlotComponent<Props> | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props> = {},
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, elementType } = options;
+
+  if (value === null || (value === undefined && !required)) {
+    return undefined;
+  }
+  const metadata: SlotComponentMetadata<Props> = isSlot<Props>(value) ? value[SLOT_COMPONENT_METADATA_SYMBOL] : {};
+
+  if (elementType) {
+    metadata.elementType = elementType;
+  }
```

Comment:
AFAIR we discussed to make `elementType` mandatory. If so, we don't need a condition there.

Reply:
It looks like it's probably optional in order to reuse this function as `resolveShorthand`, which doesn't add elementType. That's a reasonable goal, but we'll definitely want the `slotFromShorthand` function to _require_ elementType.

Also `resolveShorthand` _must_ return a different type from `slotFromShorthand`, in order to prevent a dev from assigning the result of `resolveShorthand` to a slot, since it would be missing elementType.

You could possibly swap the dependencies: have `resolveShorthand` do most of the work, then `slot` would just add `elementType` metadata. Or alternatively they could both call an internal helper function.